feat(watcher): Debounce autoWatchBatchDelay

danielcompton opened the pull request:
I'm opening this up for feedback early, to make sure I'm on the right track here. My intention is to change the behaviour of `autoWatchBatchDelay` to debounce file change events, and not run the tests until at least `autoWatchBatchDelay` ms have elapsed since the last file change.

Closes #2331

dignifiedquire commented:
Thanks @danielcompton, I agree this change makes sense. There are probably some tests that need updating after this change.

danielcompton commented:
Yeah there's some tests that fail, and I still need to verify the behaviour is correct.

danielcompton commented:
(Probably also adding more tests)

wesleycho commented:
I think all that is needed here is to just adjust the delay in the tests to bounce it back a little further.

danielcompton commented:
Thanks for the tip, you're right that did fix those tests. 

I modified the `[batch interval](https://github.com/karma-runner/karma/pull/2562/files#diff-a9b72b9949983d5d75e4e99605949cd9R744)` test to test the debouncing, and it seems like that isn't working correctly. I'm not sure if this is because the tests need to mock more of lodash, or if there is something else going on? I'm at about the limits of my JS knowledge, if you can take a quick look to see if there's something obvious I'm not doing, I'd appreciate it :)

wesleycho commented on lib/file-list.js:
```diff
@@ -71,14 +71,14 @@ var List = function (patterns, excludes, emitter, preprocess, batchInterval) {
     self._emitter.emit('file_list_modified', self.files)
   }
-  var throttledEmit = _.throttle(emit, self._batchInterval, {leading: false})
+  var debouncedEmit = _.debounce(emit, self._batchInterval, {leading: false})
```
I don't think `{leading: false}` needs to be passed into the function here - debouncing never fires the leading call.

danielcompton replied:
Thanks, fixed.

wesleycho commented:
Not sure if this helps you, but in another PR for upgrading dependencies (including lodash), I had to make this adjustment: https://github.com/wesleycho/karma/blob/c80a3161dd4c345be2ec00c6691811b5dc9b2a5d/test/unit/file-list.spec.js#L770-L805

danielcompton commented:
Phew! I think I've got it all.

In the [old version](https://github.com/karma-runner/karma/blob/2a847c250bb62134d87f5230d97be8483d4a13cf/test/unit/file-list.spec.js#L730-L750), the tests were passing, but they weren't testing quite what it looked like. `list.removeFile`, `list.changeFile, and `list.addFile` all return promises and call `self._emitModified` when a file modification is accepted as valid. 

There is an important difference in behaviour between `list.removeFile` and the other too though. Due to the promise structure of `list.removeFile`, `self._emitModified` is called synchronously. The test code was calling all of these functions one after another without waiting on the result. The only function call that had any effect on the debouncing function running while inside the test scope were the `list.removeFile` calls. You can verify this by commenting out the two `list.removeFile` calls, and watch the test fail.

```
        list.changeFile('/some/b.js')
        // list.removeFile('/some/a.js') // /some/b.js, /a.txt
        // list.removeFile('/a.txt')     // /some/b.js
        list.addFile('/a.txt')        // /some/b.js, /a.txt
        list.addFile('/some/0.js')    // /some/0.js, /some/b.js, /a.txt

        clock.tick(99)
        expect(modified).to.not.have.been.called

        clock.tick(2)
        expect(modified).to.have.been.calledOnce // This will fail
```

In my tests, I extended this test with more calls to modify and change files to extend the debouncing period. However because the promises weren't being waited on, they never ran until after the test block had finished, leading to very confusing behaviour where removing a file makes the tests work, but changing or adding a file doesn't.

I've converted the tests to wait on the promises and added more and they all pass.

danielcompton commented:
This is ready for review now.

danielcompton commented:
@wesleycho would you able to take another review please?

dignifiedquire commented:
Thanks a lot for getting through this and especially updating the tests to correctly test what they should test!
